sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +969,304 @@ cdef class MSE(RegressionCriterion):
+        # left should have all elements.
+        for k in range(self.n_outputs):
+            # if right has no elements, it's already reset
+            for i in range((<MedianHeap> right_child_heaps[k]).size()):
+                # remove everything from right and put it into left
+                (<MedianHeap> right_child_heaps[k]).pop(&popped)
+                (<MedianHeap> left_child_heaps[k]).push(popped)
+
+    cdef void update(self, SIZE_t new_pos) nogil:
+        """Updated statistics by moving samples[pos:new_pos] to the left."""
+
+        cdef double* sample_weight = self.sample_weight
+        cdef SIZE_t* samples = self.samples
+
+        cdef void** left_child_heaps = <void**> self.left_child_heaps.data
+        cdef void** right_child_heaps = <void**> self.right_child_heaps.data
+
+        cdef DOUBLE_t* y = self.y
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef SIZE_t k
+        cdef DOUBLE_t w = 1.0
+        cdef DOUBLE_t y_ik
+        cdef DOUBLE_t w_y_ik
+
+        # Update statistics up to new_pos
+        #
+        # We are going to update right_child_heaps and left_child_heaps
+        # from the direction that require the least amount of
+        # computations, i.e. from pos to new_pos or from end to new_pos.
+
+        if (new_pos - pos) <= (end - new_pos):
+            for p in range(pos, new_pos):
+                i = samples[p]
+
+                if sample_weight != NULL:
+                    w = sample_weight[i]
+
+                for k in range(self.n_outputs):
+                    y_ik = y[i * self.y_stride + k]
+                    w_y_ik = w * y_ik
+
+                    # remove w_y_ik from right and add to left
+                    (<MedianHeap> right_child_heaps[k]).remove(w_y_ik)
+                    (<MedianHeap> left_child_heaps[k]).push(w_y_ik)
+
+                self.weighted_n_left += w
+        else:
+            self.reverse_reset()
+
+            for p in range(end - 1, new_pos - 1, -1):
+                i = samples[p]
+
+                if sample_weight != NULL:
+                    w = sample_weight[i]
+
+                for k in range(self.n_outputs):
+                    y_ik = y[i * self.y_stride + k]
+                    w_y_ik = w * y_ik
+
+                    # remove w_y_ik from left and add to right
+                    (<MedianHeap> left_child_heaps[k]).remove(w_y_ik)
+                    (<MedianHeap> right_child_heaps[k]).push(w_y_ik)
+
+                self.weighted_n_left -= w
+
+        self.weighted_n_right = (self.weighted_n_node_samples - 
+                                 self.weighted_n_left)
+        self.pos = new_pos
+
+    cdef void node_value(self, double* dest) nogil:
+        """Computes the node value of samples[start:end] into dest."""
+
+        cdef SIZE_t k
+        for k in range(self.n_outputs):
+            dest[k] = self.node_medians[k]
```

Comment:
Yes, you are correct.

On Fri, Jun 24, 2016, 1:52 PM Gilles Louppe <notifications@github.com>
wrote:

> In sklearn/tree/_criterion.pyx
> <https://github.com/scikit-learn/scikit-learn/pull/6667#discussion_r68356215>
> :
>
> > +                    # remove w_y_ik from left and add to right
> > +                    (<MedianHeap> left_child_heaps[k]).remove(w_y_ik)
> > +                    (<MedianHeap> right_child_heaps[k]).push(w_y_ik)
> > +
> > +                self.weighted_n_left -= w
> > +
> > +        self.weighted_n_right = (self.weighted_n_node_samples -
> > +                                 self.weighted_n_left)
> > +        self.pos = new_pos
> > +
> > +    cdef void node_value(self, double* dest) nogil:
> > +        """Computes the node value of samples[start:end] into dest."""
> > +
> > +        cdef SIZE_t k
> > +        for k in range(self.n_outputs):
> > +            dest[k] = self.node_medians[k]
>
> So this is not correct either when sample weights are given?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6667/files/20d6107c7ebfba1b96d483cfacb97a9ac304b3cb#r68356215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AG72X8TD5C_HFuw5WwP2Q5yc5CepBGSXks5qO3CRgaJpZM4IIwnV>
> .
>